daemon/containerd/service.go

```diff
@@ -127,21 +128,64 @@ func (i *ImageService) StorageDriver() string {
 	return i.snapshotter
 }
 
+// ImageDiskUsage returns the number of bytes used by content and layer stores
+// called from disk_usage.go
+func (i *ImageService) ImageDiskUsage(ctx context.Context) (int64, error) {
+	diskUsage, err := i.layerDiskUsage(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	// Include the size of content size from the images.
+	imgs, err := i.images.List(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	visitedImages := make(map[digest.Digest]struct{})
+	for _, img := range imgs {
+		if _, ok := visitedImages[img.Target.Digest]; ok {
+			continue
+		}
+
+		if err = i.walkReachableImageManifests(ctx, img, func(imageManifest *ImageManifest) error {
+			dgst := imageManifest.Target().Digest
+			if _, ok := visitedImages[dgst]; ok {
+				return nil
+			}
+			visitedImages[dgst] = struct{}{}
+
+			size, err := imageManifest.PresentContentSize(ctx)
+			if err != nil {
+				return err
+			}
+			diskUsage += size
+			return nil
+		}); err != nil {
+			return 0, err
+		}
+		visitedImages[img.Target.Digest] = struct{}{}
```

Comment:
This will still count the same content twice if it's shared by multiple images.

For example, let's say I have `ubuntu` image and `my-ubuntu` built off:
```dockerfile
FROM ubuntu
RUN apt-get update && apt-get -y install fish
```

`my-ubuntu` will have two layers - one that is shared with `ubuntu` (let's call it `X`) and the second one that would be the delta produced by the `RUN apt-get ...` (let's call it `Y`).

Currently the code would count the `X` layer twice, which would not correspond to the actual disk usage because the layer blob is actually stored only once.

This needs to traverse each child content of the image and then only count the usage of the unique blobs.

Reply:
Would this be doable by just using `walkPresentChildren` on each of the images directly? It seems like that's what the internals of `PresentContentSize` is doing so I'm not sure if I need to call `walkReachableImageManifests` at all.